common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -504,6 +507,52 @@ public static void setMemory(long address, long bytes, byte value) {
         PlatformDependent0.setMemory(address, bytes, value);
     }
 
+    public static ByteBuffer allocateDirectNoCleaner(int capacity) {
+        if (DIRECT_MEMORY_COUNTER != null) {
+            for (;;) {
+                long usedMemory = DIRECT_MEMORY_COUNTER.get();
+                long newUsedMemory = usedMemory + capacity;
+                if (newUsedMemory > maxDirectMemory()) {
+                    throw new OutOfDirectMemoryError("failed to allocate " + capacity
+                            + " byte(s) of direct memory (used: " + usedMemory + ", max: " + maxDirectMemory() + ')');
+                }
+                if (DIRECT_MEMORY_COUNTER.compareAndSet(usedMemory, newUsedMemory)) {
+                    break;
+                }
+            }
+        }
+        try {
+            return PlatformDependent0.allocateDirectNoCleaner(capacity);
+        } catch (Throwable e) {
+            if (DIRECT_MEMORY_COUNTER != null) {
+                DIRECT_MEMORY_COUNTER.addAndGet(-capacity);
+            }
+            throwException(e);
+            return null;
+        }
+    }
+
+    public static void freeDirectNoCleaner(ByteBuffer buffer) {
```

Comment:
This is public. So a user could call this method without calling `allocateDirectNoCleaner`. At a minimum we should add comments indicating this method should only be called on buffers allocated via `allocateDirectNoCleaner` otherwise the direct memory counter will be invalid.